Add support for reading and writing <volume> tags (#41)
scalableminds/wknml: docs/wknml.md

Contains common metadata for NML files

**Notes**:

>   Setting a task or user bounding boxes will cause wK to 1) render these visually and 2) prevent data loading from outside them.

Author: Well, it already is super wK-specific is it not? Tasks are very wK. NMLs are very wK.

>   Setting a task or user bounding boxes will cause wK to 1) render these visually and 2) prevent data loading from outside them.

Reviewer: Yes, I see your point. I just wasn't sure whether this information could be better suited for the wK tasks documentation. I don't have a strong opinion on this, your call.

  

**Attributes**:

- `name` _str_ - Name of a dataset that the annotation is based on. Will cause wK to open the given skeleton annotation with the referenced dataset.

> - `scale` _Vector3_ - Voxel scale of the referenced dataset in nanometers.

Author: Yes, it would be great to automatically document/export the custom type definition as well.

  wknml.write_nml(f, nml)
  ```

<a name="wknml.nml_generation"></a>
# wknml.nml\_generation

<a name="wknml.nml_generation.random_color_rgba"></a>
#### random\_color\_rgba

```python
random_color_rgba() -> Tuple[float, float, float, float]
```

A utility to generate a new random RGBA color.

<a name="wknml.nml_generation.discard_children_hierarchy"></a>
#### discard\_children\_hierarchy

```python
discard_children_hierarchy(groups: List[Group]) -> List[Group]
```

A utility to flatten the group structure. All sub-groups will become top-level items.

<a name="wknml.nml_generation.globalize_tree_ids"></a>
#### globalize\_tree\_ids

```python
globalize_tree_ids(group_dict: Dict[str, List[nx.Graph]])
```

A utility to in-place re-assign new and globally unqiue IDs to all Tree objects. Starts with ID 1

**Arguments**:

> - `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of tree as NetworkX graph objects

Reviewer: ```suggestion
- `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of trees as NetworkX graph objects
```

<a name="wknml.nml_generation.globalize_node_ids"></a>
#### globalize\_node\_ids

```python
globalize_node_ids(group_dict: Dict[str, List[nx.Graph]])
```

A utility to in-place re-assign new and globally unqiue IDs to all Node objects. Edges are updated accordingly. Starts with ID 1.

> Note: Does not update any `Comment`s or `BranchPoint`s referencing these nodes.

Reviewer: In this state (where the trees are nx.Graphs), branchpoints and comments are indicated by a `branchpoint` or `comment` attribute at the respective node, so there is no need to update any IDs. The separate branchpoints and comments lists are only generated from the nx.Graphs, later.

**Arguments**:

> - `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of tree as NetworkX graph objects

Reviewer: ```suggestion
- `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of trees as NetworkX graph objects
```

<a name="wknml.nml_generation.generate_nml"></a>
#### generate\_nml

```python
generate_nml(tree_dict: Union[List[nx.Graph], Dict[str, List[nx.Graph]]], parameters: Dict[str, Any] = {}, globalize_ids: bool = True, volume: Optional[Dict[str, Any]] = None) -> NML
```

A utility to convert a [NetworkX graph object](https://networkx.org/) into wK NML skeleton annotation object. Accepts both a simple list of multiple skeletons/trees or a dictionary grouping skeleton inputs.

**Arguments**:

> - `tree_dict` _Union[List[nx.Graph], Dict[str, List[nx.Graph]]]_ - A list of wK tree-like structures as NetworkX graphs or a dictionary of group names and same list of NetworkX tree objects.

Reviewer: ```suggestion
- `tree_dict` _Union[List[nx.Graph], Dict[str, List[nx.Graph]]]_ - A list of wK tree-like structures as NetworkX graphs or a dictionary of group names and lists of NetworkX tree objects.
```

- `parameters` _Dict[str, Any]_ - A dictionary representation of the skeleton annotation metadata. See `NMLParameters` for accepted attributes.
- `globalize_ids` _bool = True_ - An option to re-assign new, globally unique IDs to all skeletons. Default: `True`

> - `volume` _Optional[Dict[str, Any]] = None_ - A dictionary representation of a reference to wK a volume annotation. See `Volume` object for attributes.

Reviewer: ```suggestion
- `volume` _Optional[Dict[str, Any]] = None_ - A dictionary representation of a reference to a wK volume annotation. See `Volume` object for attributes.
```

  

**Returns**:

- `nml` _NML_ - A wK NML skeleton annotation object

>   1. A dictionary with group names as keys and list of all respective NML trees as values
>   2. A dictionary representation of the NML metadata parameters

Reviewer: ```suggestion
```

Looks like this is a leftover as these are not returned by `generate_nml`.

<a name="wknml.nml_generation.generate_graph"></a>
#### generate\_graph

```python
generate_graph(nml: NML) -> Tuple[Dict[str, List[nx.Graph]], Dict[Text, any]]
```

> A utility to convert wK NML object into a [NetworkX graph object](https://networkx.org/). Skeletons/Trees are grouped by the provided groups in the NML file.

Reviewer: ```suggestion
A utility to convert a wK NML object into a [NetworkX graph object](https://networkx.org/). Skeletons/Trees are grouped by the provided groups in the NML file.
```

**Arguments**:

- `nml` _NML_ - A wK NML skeleton annotation object
  

**Returns**:

  A tuple consisting of:
  1. A dictionary with group names as keys and list of all respective NML trees as values
  2. A dictionary representation of the NML metadata parameters. See `NMLParameters` for attributes.

<a name="wknml.nml_generation.nml_tree_to_graph"></a>
#### nml\_tree\_to\_graph

```python
nml_tree_to_graph(tree: Tree) -> nx.Graph
```

A utility to convert a single wK Tree object into a [NetworkX graph object](https://networkx.org/).

<a name="wknml.nml_generation.extract_nodes_and_edges_from_graph"></a>
#### extract\_nodes\_and\_edges\_from\_graph

```python
extract_nodes_and_edges_from_graph(graph: nx.Graph) -> Tuple[List[Node], List[Edge]]
```

> A utility to convert a single [NetworkX graph object](https://networkx.org/) into a list of `Node`objects and `Edge` objects.

Reviewer: ```suggestion
A utility to convert a single [NetworkX graph object](https://networkx.org/) into a list of `Node` objects and `Edge` objects.
```

Return

>     Tuple[List[Node], List[Edge]]: A tuple contain both all nodes and edges

Reviewer: ```suggestion
    Tuple[List[Node], List[Edge]]: A tuple containing all nodes and all edges
```
